### PR TITLE
feat: add ability to prefix all exports with the schema name

### PR DESCRIPTION
**Pull request by jamesone**
When you are working with multiple schemas with potentially similar table names it can get a bit messy and hard to manage. This solution aims to solve it.

### CODE REVIEW

**Comment by jamesone** [comment]
@danvk I would be willing to help maintain this library (if that is something you would want help with)

File: README.md [hunk 1]
```diff
@@ -142,6 +142,47 @@ export interface ProductMetadata {
...
+  created_at: Date;
+}
+export interface MaxiProductInput {
+  id?: string;
+  name: string;
+  description: string;
+  created_at?: Date;
+}
+const maxi_product = {
+  tableName: 'product',
```
Comment (danvk) [inline README.md:169]: Should this be `maxi.product`?

File: src/options.ts [hunk 2]
```diff
@@ -13,6 +14,10 @@ export type OptionValues = {
   writeHeader?: boolean; // write schemats description header
   /** Import types for jsdoc-tagged JSON columns from this path */
   jsonTypesFile?: string;
+  /**
```
Comment (danvk) [inline src/options.ts:17]: nit: for very short jsdoc comments, I prefer the single-line form:

```ts
/** Prefix the tablenames with the schema name. */
```

File: src/schemaPostgres.ts [hunk 3]
```diff
@@ -215,6 +218,10 @@ export class PostgresDatabase {
         'WHERE table_schema = $1 ' +
         'GROUP BY table_name ORDER BY lower(table_name)',
       [schemaName],
+      /**
+       * Customisations added:
```
Comment (danvk) [inline src/schemaPostgres.ts:222]: I'm not sure what this is referring to? Is there a change here?

File: src/cli.ts [hunk 4]
```diff
@@ -68,6 +68,10 @@ const argv = yargs(hideBin(process.argv))
             'See https://github.com/brianc/node-pg-types for details.',
           type: 'boolean',
         },
+        prefixWithSchemaNames: {
+          describe: 'Prefix the table name to the types',
```
Comment (danvk) [inline src/cli.ts:72]: I think you mean schema name?

File: src/schemaPostgres.ts [hunk 5]
```diff
@@ -10,6 +10,9 @@ import {
 
 const pgp = PgPromise();
 
+/**
+ * Converts a postgres data type to a typescript
```
Comment (danvk) [inline src/schemaPostgres.ts:14]: I think this is already clear from the function name :)

File: src/typescript.ts [hunk 6]
```diff
@@ -15,6 +15,17 @@ function nameIsReservedKeyword(name: string): boolean {
   return reservedKeywords.indexOf(name) !== -1;
 }
 
+/**
+ * Will determine whether the tableName should be prefixed with the schemaName
+ */
+export function getEnhancedTableName(
```
Comment (danvk) [inline src/typescript.ts:21]: I don't love "enhanced" -- maybe `getTableTypeName`?

File: src/typescript.ts [hunk 7]
```diff
@@ -15,6 +15,17 @@ function nameIsReservedKeyword(name: string): boolean {
...
+  tableName: string,
+  schema: string,
+  prefixWithSchemaNames: boolean,
+) {
+  return `${prefixWithSchemaNames ? `${schema}_` : ''}${tableName}`;
```
Comment (danvk) [inline src/typescript.ts:26]: nit: avoid nesting template strings, I think this is clearer as:

```ts
return (prefixWithSchemaNames ? `${schema}_` : '') + tableName;
```

File: README.md [hunk 8]
```diff
@@ -142,6 +142,47 @@ export interface ProductMetadata {
 Note that, on its own, TypeScript cannot enforce a schema on your `json`
 columns. For that, you'll want a tool like [postgres-json-schema][].
 
+
+### Prefix tableNames with there corresponding schemaName
+
+--prefixWithSchemaNames
```
Comment (danvk) [inline README.md:148]: Put this in backticks for code formatting

**Comment by danvk** [comment]
Hi @jamesone, thanks for the contribution, and especially for the offer to help maintain!

A few high level comments:

- There's an existing `-s` option to generate types for just one Postgres schema. You could use this to generate a different TS file for each schema and get a natural sort of namespacing that way. Would that meet your needs?
- Please add an integration test with this option set (probably to `schema-generation-test.ts`). There are some notes on how to do this in [`test/README.md`](/test/README.md).

**Review by danvk — CHANGES_REQUESTED** [review]

**Comment by jamesone** [comment]
> Hi @jamesone, thanks for the contribution, and especially for the offer to help maintain!
> 
> 
> 
> A few high level comments:
> 
> 
> 
> - There's an existing `-s` option to generate types for just one Postgres schema. You could use this to generate a different TS file for each schema and get a natural sort of namespacing that way. Would that meet your needs?
> 
> - Please add an integration test with this option set (probably to `schema-generation-test.ts`). There are some notes on how to do this in [`test/README.md`](/test/README.md).
> 
> 

Hey @danvk I will get back to you on these points you brought up and the requested changes as soon as I can! Something has popped up this week and I'm unable to work. Appreciate you reviewing the PR.

File: src/schemaPostgres.ts (hunk 3 above)
Comment (jamesone) [reply src/schemaPostgres.ts:222]: removing 👍 (was an old change I added but forgot to rm comment)

File: src/cli.ts (hunk 4 above)
Comment (jamesone) [reply src/cli.ts:72]: Changed to: 
```ts
prefixWithSchemaNames: {
          describe:
            'Prefix the schema name to the table name. ' +
            ' All exports will be prefixed with the schema name.',
          type: 'boolean',
        },
```

File: src/schemaPostgres.ts (hunk 5 above)
Comment (jamesone) [reply src/schemaPostgres.ts:14]: Agreed!

File: src/typescript.ts (hunk 6 above)
Comment (jamesone) [reply src/typescript.ts:21]: Yep I like that better

File: src/typescript.ts (hunk 7 above)
Comment (jamesone) [reply src/typescript.ts:26]: Nice! I like that solution

**Comment by jamesone** [comment]
Hey, @danvk I made the requested changes and acted on your comments. The integration test is in too. 

> Also to answer your question: here's an existing -s option to generate types for just one Postgres schema. You could use this to generate a different TS file for each schema and get a natural sort of namespacing that way. Would that meet your needs?
> I did know about this and utilized it but it didn't meet the needs of my project.

File: test/integration/__snapshots__/integration.test.ts.snap [hunk 9]
```diff
@@ -429,3 +429,44 @@ export const tables = {
...
+  name: string;
+  pronoun: string | null;
+}
+export interface PublicUsersInput {
+  id?: string;
+  name: string;
+  pronoun?: string | null;
+}
+const public_Users = {
+  tableName: 'public.Users',
```
Comment (danvk) [inline test/integration/__snapshots__/integration.test.ts.snap:453]: This should be `public.users` (lowercase `u`). (My suggestion above should fix this.)

File: test/integration/__snapshots__/integration.test.ts.snap [hunk 10]
```diff
@@ -429,3 +429,44 @@ export const tables = {
...
+  id: string;
+  name: string;
+  pronoun: string | null;
+}
+export interface PublicUsersInput {
+  id?: string;
+  name: string;
+  pronoun?: string | null;
+}
+const public_Users = {
```
Comment (danvk) [inline test/integration/__snapshots__/integration.test.ts.snap:452]: This variable name looks a bit strange. Can we make it either `public_users` or `publicUsers`? The former feels more consistent with the existing un-prefixed names.

**Review by danvk — APPROVED** [review]
A few more suggestions, mostly small things. Thanks for adding the tests! This should be good to merge on the next go-round.

File: src/typescript.ts [hunk 11]
```diff
@@ -94,23 +115,35 @@ export function generateTableInterface(
...
-      // Table ${tableName}
+      // Table ${enhancedTableName}
       ${jsdoc} export interface ${camelTableName} {
         ${selectableMembers}}
       ${jsdoc} export interface ${camelTableName}Input {
         ${insertableMembers}}
       const ${normalizedTableName} = {
-        tableName: '${tableName}',
+        tableName: '${
+          options.options.prefixWithSchemaNames
```
[the remark this answers is not in the document]
Comment (jamesone) [reply src/typescript.ts:143]: Good pickup 👍

File: src/index.ts [hunk 12]
```diff
@@ -103,8 +104,15 @@ export async function typescriptOfSchema(
   }
 
   const tableNames = tables.map(t =>
-    normalizeName(optionsObject.transformTypeName(t)),
+    normalizeName(
+      getTableTypeName(
+        optionsObject.transformTypeName(t),
+        schema as string,
```
[the remark this answers is not in the document]
Comment (jamesone) [reply src/index.ts:110]: yeah good point! I like the `inSchema` solution

File: src/index.ts [hunk 13]
```diff
@@ -103,8 +101,9 @@ export async function typescriptOfSchema(
   }
 
   const tableNames = tables.map(t =>
-    normalizeName(optionsObject.transformTypeName(t)),
+    normalizeName(getTableTypeName(t, schema, !!options.prefixWithSchemaNames)),
```
Comment (jamesone) [inline src/index.ts:104]: @danvk I believe transformTypeName has been used wrong in places where it's given the table name.  

I found this out when fixing the implementation of this: https://github.com/danvk/pg-to-ts/pull/14#discussion_r839645868

For some reason the code was transforming the table name to camelCase IF the camelCase option was set. From what I can see the camelCase option is for transforming the column names. I think my implementation can stay if we don't want the table names to be exported as camelCase

File: test/integration/__snapshots__/integration.test.ts.snap [hunk 14]
```diff
@@ -294,8 +294,8 @@ export interface UsersInput {
   jsonbArrayCol?: Json[] | null;
   timestamptzArrayCol?: Date[] | null;
 }
-const Users = {
-  tableName: 'Users',
```
Comment (jamesone) [inline test/integration/__snapshots__/integration.test.ts.snap:298]: RE my change here: https://github.com/danvk/pg-to-ts/pull/14/files/272aacaf8d5cba9a120ff44d10e89e21cd590f7f#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R104


This change I made corrects this. The camelCase option is meant to apply camelCase to the column names only, right?

Comment (jamesone) [reply test/integration/__snapshots__/integration.test.ts.snap:298]: The description of the camelCase option is: `Camel-case columns (e.g. user_id --> userId)`

Comment (jamesone) [reply test/integration/__snapshots__/integration.test.ts.snap:298]: I think it makes sense to only make the column names camelCase (like it says in the description of the camelCase option).

File: src/index.ts (hunk 13 above)
Comment (danvk) [reply src/index.ts:104]: Yes, that makes sense to me. Let's fix this in a future change.

**Review by danvk — APPROVED** [review]
Looks great! Merging…

**Comment by danvk** [comment]
Released in 4.1.0.

**Pull request merged**